# **Proyecto Final: Cadenas de Markov**

**Estadística Aplicada a la Teoría de Decisiones II**   
Maestría en Investigación de Operaciones  
Universidad Galileo  

Este cuaderno contiene el desarrollo completo del proyecto final sobre *Cadenas de Markov*, abordando los siguientes puntos:

1. **Marco Teórico**  
2. **Resolución de Problemas**  
3. **Aplicaciones Reales**  

Se han utilizado herramientas matemáticas, computacionales y de simulación para resolver los ejercicios planteados y analizar los resultados de forma clara y estructurada.

## 1. **Marco Teórico**

### Introducción

Las **Cadenas de Markov** constituyen una herramienta analítica esencial para modelar y predecir comportamientos en sistemas donde el estado futuro depende únicamente del estado presente. Su gran versatilidad permite utilizarlas en una amplia variedad de ámbitos, que van desde la economía y las finanzas hasta la ingeniería y las ciencias de la salud. En las secciones siguientes se profundizará en los fundamentos teóricos de las Cadenas de Markov, sus principales características, y las condiciones que aseguran la existencia de un estado estacionario.

### 1.1 **¿Qué es una Cadena de Markov?**

Una **Cadena de Markov** es un proceso estocástico (es decir, un proceso que involucra aleatoriedad) en el cual el estado futuro depende únicamente del estado presente y no de la secuencia de estados anteriores. En términos matemáticos:

$$
P\bigl(X_{n+1} = x \,\big|\ X_n = x_n,\ X_{n-1} = x_{n-1}, \dots, X_0 = x_0 \bigr) \;=\; P\bigl(X_{n+1} = x \,\big|\ X_n = x_n \bigr).
$$

Esta propiedad, conocida como **propiedad de Markov** o **memoria limitada**, permite simplificar en gran medida el análisis del proceso, ya que las probabilidades de transición únicamente necesitan considerar el estado actual.


### 1.2 **Diferencia entre Cadenas de Tiempo Discreto y Tiempo Continuo**

- **Cadenas de Markov en tiempo discreto**:  
  En este caso, el tiempo avanza en pasos o intervalos discretos (por ejemplo, \( n = 0, 1, 2, \dots \)). La transición de un estado a otro se produce en cada paso y viene dada por una **matriz de transición**. Es el tipo de cadena más común en muchos modelos de procesos secuenciales (como días, meses, iteraciones, etc.).

- **Cadenas de Markov en tiempo continuo**:  
  En estas cadenas, el proceso evoluciona de manera continua en el tiempo, y las transiciones pueden ocurrir en cualquier instante. Por lo general, se modelan mediante tasas de transición (matriz generador infinitesimal) y utilizan distribuciones continuas (a menudo, la distribución exponencial) para describir el tiempo que se permanece en cada estado antes de dar el salto a otro.

La principal diferencia está en **cómo se mide el tiempo y cómo se define el cambio de estado**, lo que también impacta en las herramientas matemáticas que se utilizan para analizarlas.

### 1.3 **Conceptos Fundamentales**

#### 1.3.1 **Espacio de Estados**

El **espacio de estados** es el conjunto de todos los valores o condiciones posibles que puede tomar el proceso. Por ejemplo, en un modelo sobre estados emocionales, el espacio podría ser \\(\{ \text{Feliz, Neutro, Triste}\}\\). En un modelo de clientes, podría ser \\(\{\text{Fiel, Intermitente, Perdido}\}\\).

#### 1.3.2 **Matriz de Transición**

Una **matriz de transición**, denotada comúnmente como \\(P\\), contiene las probabilidades de pasar de cada estado a cada uno de los posibles estados en un solo paso de tiempo. Cada elemento \\(p_{ij}\\) de la matriz \\(P\\) representa la probabilidad de ir del estado \\(i\\) al estado \\(j\\):

$$
P \;=\;
\begin{pmatrix}
p_{11} & p_{12} & \cdots & p_{1m} \\
p_{21} & p_{22} & \cdots & p_{2m} \\
\vdots & \vdots & \ddots & \vdots \\
p_{m1} & p_{m2} & \cdots & p_{mm}
\end{pmatrix},
$$

donde cada fila suma 1.

#### 1.3.3 **Probabilidad de Transición en \\(n\\) Pasos**

La **probabilidad de transición en \\(n\\) pasos** es la probabilidad de que el proceso pase del estado \\(i\\) al estado \\(j\\) después de \\(n\\) pasos. Se denota frecuentemente como \\(p_{ij}^{(n)}\\) y puede calcularse a partir de la \\(n\\)-ésima potencia de la matriz de transición:

$$
P^{(n)} \;=\; P^n \;=\;
\underbrace{P \times P \times \cdots \times P}_{n\text{ veces}}.
$$

#### 1.3.4 **Estado Estacionario**

Un **estado estacionario** (o **distribución estacionaria**) es un vector de probabilidades \\(\pi\\) tal que, una vez alcanzado, describe la proporción de tiempo que el proceso pasará en cada estado a largo plazo. Cumple la condición:

$$
\pi \, P \;=\; \pi,
$$

donde \\(\pi\\) es un vector fila cuyas entradas son las probabilidades de encontrarse en cada estado. Además, debe cumplirse:

$$
\sum_{i} \pi_i \;=\; 1
\quad \text{con} \quad 
\pi_i \geq 0.
$$


### 1.4 **Condiciones para la Existencia de un Estado Estacionario**

Para que exista un **estado estacionario** único en una cadena de Markov en tiempo discreto, se requieren dos condiciones principales:

1. **Irreducibilidad**: La cadena es irreducible si es posible llegar de cualquier estado a cualquier otro estado (no necesariamente en un solo paso, sino en un número finito de pasos).

2. **Aperiodicidad**: La cadena es aperiódica si cada estado no se visita únicamente en intervalos regulares. En otras palabras, el máximo común divisor de los posibles tiempos de retorno a un estado es 1.

3. **Cadena de tiempo discreto y espacio de estados finito**: Cuando la cadena es de tiempo discreto y tiene un número finito de estados, estas condiciones son suficientes para garantizar la existencia de un **único vector estacionario** al que la cadena converge, sin importar el estado inicial.

**Resumen:**

Si una cadena de Markov es **irreducible** y **aperiódica**, y tiene **espacio de estados finito**, entonces:

- Existe un vector estacionario $\pi$
- El sistema converge a $\pi$ con el tiempo:  
  $$ \lim_{n \to \infty} P^n = \begin{bmatrix} \pi \\ \pi \\ \vdots \\ \pi \end{bmatrix} $$
- La distribución de probabilidad a largo plazo es independiente del estado inicial


Bajo estas condiciones, se garantiza la existencia de una única distribución estacionaria que describe el comportamiento a largo plazo de la cadena.




## 2. **Resolución de Problemas**

En esta sección se presentan tres ejercicios de aplicación directa de las Cadenas de Markov. Cada ejercicio aborda un escenario distinto, ilustrando los procedimientos de análisis a corto plazo, el cálculo de estados estacionarios y la modelación de sistemas.


### 2.1 **Ejercicio 1: Probabilidades a corto plazo**

Una persona puede estar en tres estados emocionales cada día:

- **Feliz (H)**
- **Neutra (N)**
- **Triste (T)**

Las transiciones diarias están dadas por la siguiente matriz:

$$
P = 
\begin{pmatrix}
0.85 & 0.10 & 0.05 \\
0.40 & 0.40 & 0.20 \\
0.10 & 0.30 & 0.60
\end{pmatrix}.
$$

Si la persona empieza el lunes en estado **feliz**, calcule la probabilidad de que esté en cada uno de los estados el **miércoles** (es decir, dos pasos después).

**Objetivo**  
- Determinar las probabilidades a corto plazo (dos pasos) para cada estado, dada una condición inicial.
- Para ello, se debe utilizar la matriz $(P)$ y calcular $(P^2)$.

**Puntos Clave**  
- Definir correctamente el espacio de estados.
- Utilizar la matriz de transición para calcular $(P^2)$.
- Multiplicar la distribución inicial (100% en Feliz) por $(P^2)$.


**Paso 1: Definir la matriz de transición y el vector de estado inicial**

Dado que la persona está **feliz** el lunes, el vector de estado inicial es:

$$
\pi_0 = \begin{bmatrix} 1 & 0 & 0 \end{bmatrix}
$$

**Paso 2: Calcular $\pi_2 = \pi_0 \cdot P^2$**

Este cálculo nos dará la probabilidad de que esté feliz, neutral o triste el miércoles.

In [126]:
import numpy as np
import pandas as pd

# Matriz de transición
P = np.array([
    [0.85, 0.10, 0.05],
    [0.40, 0.40, 0.20],
    [0.10, 0.30, 0.60]
])

# Vector de estado inicial: feliz
pi_0 = np.array([1, 0, 0])

# Probabilidades al miércoles (dos pasos)
pi_2 = pi_0 @ np.linalg.matrix_power(P, 2)

# Mostrar resultados
estados = ['Feliz (H)', 'Neutra (N)', 'Triste (T)']
df_resultado = pd.DataFrame({'Estado': estados, 'Probabilidad al Miércoles': pi_2})

# Imprimir tabla
print(df_resultado)



       Estado  Probabilidad al Miércoles
0   Feliz (H)                     0.7675
1  Neutra (N)                     0.1400
2  Triste (T)                     0.0925


**Interpretación de resultados**

El vector resultante $\pi_2$ nos indica la probabilidad de que la persona se encuentre en cada uno de los estados emocionales el **miércoles**, habiendo comenzado **feliz** el lunes.

Los valores obtenidos reflejan lo siguiente:

- **Feliz (H):** Esta es la probabilidad más alta, lo cual es consistente con que el modelo tiene una fuerte tendencia a permanecer en este estado (0.85 en la matriz).
- **Neutra (N):** La probabilidad intermedia sugiere que, aunque es menos común el cambio a neutral, puede ocurrir en dos pasos con cierta frecuencia.
- **Triste (T):** La probabilidad más baja, indicando que es menos probable que una persona inicialmente feliz llegue a estar triste en tan corto plazo.

En resumen, el modelo predice una alta persistencia en el estado **feliz** en el corto plazo, con una transición moderada hacia la neutralidad y baja hacia la tristeza.


### 2.2 **Ejercicio 2: Estado estacionario**

Una empresa clasifica a sus clientes en tres categorías:

- Fieles (F)  
- Intermitentes (I)  
- Perdidos (P)  

La siguiente matriz de transición mensual describe cómo se mueven los clientes entre estas categorías:

$$
P = \begin{bmatrix}
0.85 & 0.10 & 0.05 \\
0.40 & 0.40 & 0.20 \\
0.10 & 0.30 & 0.60 \\
\end{bmatrix}
$$

Se pide:

a) Determinar si existe un estado estacionario.  
b) Si existe, calcular el vector estacionario.  
c) Interpretar el significado de este vector para la empresa.


#### 2.2.a **Determinar si existe un estado estacionario.**

**Análisis teórico**

Para que exista un estado estacionario único en una cadena de Markov con espacio de estados finito y tiempo discreto, deben cumplirse dos condiciones:

1. **Irreducibilidad:** Desde cualquier estado se puede alcanzar cualquier otro (aunque sea en varios pasos).
2. **Aperiodicidad:** No existen ciclos fijos que obliguen al sistema a volver a un estado solo en múltiplos de cierto número.


**Análisis de la matriz**

- Todas las entradas de la matriz de transición son **positivas o no nulas**, lo cual indica que eventualmente se puede transitar de un estado a otro.
- Además, dado que algunos estados pueden transicionar hacia sí mismos (por ejemplo, $P_{11} = 0.85$), la cadena **no está atrapada en ciclos periódicos**.


**Conclusión**

La cadena de Markov definida por la matriz:

$$
P = \begin{bmatrix}
0.85 & 0.10 & 0.05 \\
0.40 & 0.40 & 0.20 \\
0.10 & 0.30 & 0.60 \\
\end{bmatrix}
$$

es **irreducible** y **aperiódica**, por lo tanto:

> **Sí existe un estado estacionario único al cual la cadena converge.**

#### 2.2.b **Si existe, calcule el vector estacionario.**

Método para hallar el vector estacionario

El vector estacionario $\pi$ es aquel que cumple:

$$
\pi P = \pi
$$

con la condición:

$$
\pi_1 + \pi_2 + \pi_3 = 1
$$

Para resolver este sistema, se puede:

1. Restar $\pi P = \pi$ como $\pi(P - I) = 0$
2. Reescribir el sistema como un conjunto de ecuaciones lineales.
3. Agregar la ecuación de normalización: $\sum \pi_i = 1$

El vector resultante $\pi = [\pi_1, \pi_2, \pi_3]$ representa las proporciones de clientes en cada estado a largo plazo.


In [130]:
import numpy as np

# Matriz de transición
P = np.array([
    [0.85, 0.10, 0.05],
    [0.40, 0.40, 0.20],
    [0.10, 0.30, 0.60]
])

# Transpuesta de P para trabajar con ecuaciones lineales: pi.T @ P = pi.T
# Restamos la identidad
A = P.T - np.eye(3)

# Agregamos la ecuación de normalización: pi_1 + pi_2 + pi_3 = 1
A = np.vstack([A, np.ones(3)])
b = np.array([0, 0, 0, 1])

# Resolver el sistema
pi = np.linalg.lstsq(A, b, rcond=None)[0]

# Mostrar el resultado
import pandas as pd
estados = ['Fieles (F)', 'Intermitentes (I)', 'Perdidos (P)']
df_pi = pd.DataFrame({'Estado': estados, 'Probabilidad estacionaria': pi})
print(df_pi)


              Estado  Probabilidad estacionaria
0         Fieles (F)                   0.631579
1  Intermitentes (I)                   0.192982
2       Perdidos (P)                   0.175439


#### 2.2.c **Interprete el significado de este vector para la empresa.**

El vector estacionario obtenido representa la distribución estable de los clientes entre los tres estados a largo plazo. Es decir, después de muchas transiciones (meses), el porcentaje de clientes en cada categoría tiende a estabilizarse en los valores obtenidos, sin importar el estado inicial.

Por ejemplo, si el vector estacionario es:

$$
\pi = [\pi_1, \pi_2, \pi_3] = [0.63, 0.19, 0.18]
$$

Esto significa que, en el largo plazo:

- Aproximadamente el **63%** de los clientes serán **fieles**.
- Un **19%** serán **intermitentes**.
- Y un **18%** estarán **perdidos**.

Para la empresa, este resultado permite:

- Estimar el tamaño estable de cada segmento de clientes.
- Identificar si la retención a largo plazo es adecuada.
- Evaluar si se requiere una estrategia para reducir la proporción de clientes perdidos o mejorar la fidelización.

Este análisis es útil para la planificación comercial, proyecciones de ingresos y diseño de estrategias de retención.


### 2.3 **Ejercicio 3: Modelación de un sistema**

Una máquina puede estar en uno de los siguientes estados:

- Funcionando bien (B)  
- Funcionando con fallas (F)  
- Detenida por reparación (R)  

Cada semana, la máquina cambia de estado según la siguiente matriz de transición:

$$
P = \begin{bmatrix}
0.7 & 0.2 & 0.1 \\
0.3 & 0.4 & 0.3 \\
0.5 & 0.3 & 0.2 \\
\end{bmatrix}
$$

Se pide:

a) Modele el proceso como una cadena de Markov.  
b) Suponga que esta semana la máquina está en **falla**. ¿Cuál es la probabilidad de que dentro de dos semanas esté **funcionando bien**?  
c) Estime, mediante simulación o iteración, el comportamiento del sistema a largo plazo.

#### 2.3.a **Modele el proceso como una cadena de Markov.**

El sistema descrito corresponde a una **cadena de Markov de tiempo discreto**, ya que:

- El estado del sistema cambia **una vez por semana**, lo que implica una evolución en pasos discretos.
- En cada transición, la probabilidad de moverse de un estado a otro depende **únicamente del estado actual** y no del historial de estados anteriores (cumple la propiedad de Markov).

Los estados son:

- **B**: Funcionando bien  
- **F**: Funcionando con fallas  
- **R**: Detenida por reparación  

La matriz de transición que describe el comportamiento semanal del sistema es:

$$
P = \begin{bmatrix}
0.7 & 0.2 & 0.1 \\\\
0.3 & 0.4 & 0.3 \\\\
0.5 & 0.3 & 0.2 \\
\end{bmatrix}
$$

Donde la fila $i$ representa el estado actual y cada columna $j$ representa la probabilidad de pasar al estado $j$ en la próxima semana.

Este modelo permite analizar el comportamiento del sistema a corto y largo plazo, y tomar decisiones informadas sobre mantenimiento, prevención de fallas y eficiencia operativa.

#### 2.3.b **Probabilidad de estar funcionando bien en dos semanas**

Se nos indica que la máquina **actualmente está funcionando con fallas (F)**, y queremos conocer la **probabilidad de que en dos semanas esté funcionando bien (B)**. Esto equivale a calcular la probabilidad de transición en dos pasos desde el estado F al estado B, es decir:

$$
P^{(2)}_{FB}
$$

Donde:

- $P^{(2)} = P^2$, es decir, la matriz de transición al segundo paso.
- El estado F corresponde a la **segunda fila** (índice 1).
- El estado B corresponde a la **primera columna** (índice 0).

In [140]:
import numpy as np

# Matriz de transición
P = np.array([
    [0.7, 0.2, 0.1],
    [0.3, 0.4, 0.3],
    [0.5, 0.3, 0.2]
])

# Elevar la matriz al segundo paso
P2 = np.linalg.matrix_power(P, 2)

# Extraer la probabilidad de ir de estado F (fila 1) a estado B (columna 0)
prob_f_to_b = P2[1, 0]

# Mostrar el resultado
print(f"Probabilidad de que la máquina esté funcionando bien en dos semanas (iniciando en falla): {prob_f_to_b:.4f}")


Probabilidad de que la máquina esté funcionando bien en dos semanas (iniciando en falla): 0.4800


**Interpretación del resultado**

La probabilidad de que la máquina esté **funcionando bien** dentro de dos semanas, comenzando esta semana en estado **de falla**, es:

$$
P^{(2)}_{FB} = 0.48
$$

Esto significa que hay un **48% de probabilidad** de que el sistema vuelva a estar operativo en condiciones normales después de dos semanas, partiendo de un estado de falla. Este resultado indica una **probabilidad moderada de recuperación** sin intervención externa. Si bien no es altamente probable, tampoco es despreciable, lo que sugiere que el sistema tiene una dinámica donde **el estado de falla no necesariamente lleva a deterioro continuo**.

Desde un punto de vista práctico, este valor podría justificar estrategias de **monitoreo activo durante dos semanas antes de intervenir**, o establecer un umbral de espera para el mantenimiento, especialmente si los costos de reparación son altos.



#### 2.3.c **Estimar el comportamiento del sistema a largo plazo**

Para conocer cómo se comportará el sistema después de muchas semanas, se debe calcular el **vector estacionario** de la matriz de transición $P$:

$$
\pi = [\pi_B, \pi_F, \pi_R]
$$

Este vector representa la **probabilidad de que la máquina esté en cada estado en el largo plazo**, sin importar el estado inicial.

Se cumple:

$$
\pi P = \pi \quad \text{y} \quad \pi_B + \pi_F + \pi_R = 1
$$

Este cálculo se puede realizar resolviendo un sistema de ecuaciones o mediante iteración hasta convergencia.


In [138]:
import numpy as np

# Matriz de transición
P = np.array([
    [0.7, 0.2, 0.1],
    [0.3, 0.4, 0.3],
    [0.5, 0.3, 0.2]
])

# Método algebraico: resolver (P' - I)π = 0 con suma de π_i = 1
A = P.T - np.eye(3)
A = np.vstack([A, np.ones(3)])        # Agregamos ecuación π1 + π2 + π3 = 1
b = np.array([0, 0, 0, 1])            # Vector resultado

# Resolver sistema
pi = np.linalg.lstsq(A, b, rcond=None)[0]

# Mostrar resultado
import pandas as pd
estados = ['Funcionando bien (B)', 'Con fallas (F)', 'En reparación (R)']
df_pi = pd.DataFrame({'Estado': estados, 'Probabilidad a largo plazo': pi})
print(df_pi)


                 Estado  Probabilidad a largo plazo
0  Funcionando bien (B)                    0.557143
1        Con fallas (F)                    0.271429
2     En reparación (R)                    0.171429


El vector estacionario calculado es:

$$
\pi = [0.5571,\ 0.2714,\ 0.1714]
$$

Esto indica que, a largo plazo, la máquina se comportará de la siguiente manera:

- Estará **funcionando bien** aproximadamente el **55.71%** del tiempo.
- Estará **funcionando con fallas** el **27.14%** del tiempo.
- Estará **detenida por reparación** el **17.14%** del tiempo.

Este análisis permite prever la disponibilidad general del sistema en el futuro y tomar decisiones sobre mantenimiento preventivo, asignación de recursos técnicos o incluso reemplazo de maquinaria.

El hecho de que más de la mitad del tiempo la máquina esté operativa es positivo, pero también se observa que cerca del **45% del tiempo** el sistema **no estará en condiciones óptimas**, lo cual puede representar un riesgo operacional o financiero si no se gestionan adecuadamente las fallas y los tiempos de reparación.
